include/nbl/video/IGPUCommandPool.h

```diff
@@ -4,38 +4,835 @@
+        static uint32_t calc_size(const Args&...)
+        {
+            return sizeof(CRTP);
+        }
+
+    protected:
+        IFixedSizeCommand() : ICommand(calc_size()) {}
+    };
+
+    class alignas(COMMAND_SEGMENT_ALIGNMENT) CCommandSegment
+    {
+    private:
+        struct header_t
         {
-            ECF_NONE = 0x00,
-            ECF_TRANSIENT_BIT = 0x01,
-            ECF_RESET_COMMAND_BUFFER_BIT = 0x02,
-            ECF_PROTECTED_BIT = 0x04
+            core::LinearAddressAllocator<uint32_t> m_commandAllocator;
+            CCommandSegment* m_next;
+        } header;
+
+    public:
+        static inline constexpr uint32_t STORAGE_SIZE = COMMAND_SEGMENT_SIZE - core::roundUp(sizeof(header_t), alignof(ICommand));
+
+        struct Iterator
+        {
+            ICommand* m_cmd = nullptr;
+            // uint32_t m_size = 0;
+            CCommandSegment* m_segment = nullptr;
+
+            inline bool operator!=(const Iterator& other) const { return m_cmd != other.m_cmd; }
+            // TODO(achal): Return Iterator&?
+            // inline void operator++()
+            // {
+            //     m_cmd = reinterpret_cast<ICommand*>(reinterpret_cast<uint8_t*>(m_cmd) + m_size);
+            //     m_size = ;
+            // }
         };
 
-        IGPUCommandPool(core::smart_refctd_ptr<const ILogicalDevice>&& dev, core::bitflag<E_CREATE_FLAGS> _flags, uint32_t _familyIx) : IBackendObject(std::move(dev)), m_flags(_flags), m_familyIx(_familyIx) {}
+        CCommandSegment()
+        {
+            header.m_commandAllocator = core::LinearAddressAllocator<uint32_t>(nullptr, 0u, 0u, COMMAND_SEGMENT_ALIGNMENT, STORAGE_SIZE);
+            header.m_next = nullptr;
 
-        core::bitflag<E_CREATE_FLAGS> getCreationFlags() const { return m_flags; }
-        uint32_t getQueueFamilyIndex() const { return m_familyIx; }
+            wipeNextCommandSize();
+        }
 
-        // OpenGL: nullptr, because commandpool doesn't exist in GL (we might expose the internal allocator in the future)
-        // Vulkan: const VkCommandPool*
-        virtual const void* getNativeHandle() const = 0;
+        template <typename Cmd, typename... Args>
+        void* allocate(const Args&... args)
+        {
+            const uint32_t cmdSize = Cmd::calc_size(args...);
+            const auto address = header.m_commandAllocator.alloc_addr(cmdSize, alignof(Cmd));
+            if (address == decltype(header.m_commandAllocator)::invalid_address)
+                return nullptr;
 
-    protected:
-        virtual ~IGPUCommandPool() = default;
+            wipeNextCommandSize();
+
+            void* cmdMem = m_data + address;
+            return cmdMem;
+        }
+
+        inline void setNext(CCommandSegment* segment) { header.m_next = segment; }
+        inline CCommandSegment* getNext() const { return header.m_next; }
+        inline ICommand* getFirstCommand() { return reinterpret_cast<ICommand*>(m_data); } // TODO(achal): Probably remove?
```

Comment:
if you're not using it, then remove it

Reply:
definitely shouldn't be around we have `begin()` for that

Reply:
Resolved.